### PR TITLE
feat: add Completion Progress page [progression]

### PR DESCRIPTION
**Pull request by wescopeland**
> ✅ Safe to merge, supports the "beat" feature flag.

This PR adds the new Completion Progress page, which is intended to be the destination of the component built in https://github.com/RetroAchievements/RAWeb/pull/1816.

The page is accessible at `/user/{username}/progress`.

## UX Walkthrough

Screenshot 2023-09-06 at 6 17 13 PM

The Completion Progress page provides an overview of a user's entire play history on RetroAchievements. Users can see their gameplay history, milestones, awards metadata, and can filter down into specific data to see the games they're interested in at the moment.

The game list consists of rows that look like:

Screenshot 2023-09-06 at 6 18 39 PM

The date is when the user last earned an achievement for that game.

If the game has an award associated with it, the award indicator changes and a new line of copy is added to the row:

Screenshot 2023-09-06 at 6 19 45 PM

The embed at the top of the page shows counts of games by various award statuses. These counts are dynamic, changing based on which console you're filtered onto at the time. Clicking one of the counts applies its respective filter:

Screenshot 2023-09-06 at 6 20 18 PM

Many progression status filtering options are provided to choose from:

Screenshot 2023-09-06 at 6 21 51 PM

Additionally, milestones can be found in the sidebar. For now, these are just showing awards granted at a handful of landmark ordinal intervals. In the future, after we cut over to aggregated queries, it would make sense to show things such as: first achievement earned, rarest achievement earned, rarest mastery, rarest beaten game, etc.:

Screenshot 2023-09-06 at 6 23 15 PM

Like award totals, milestones also update dynamically based on your current filter criteria.

Only 100 results are shown per page. A pagination control is provided at the bottom of the screen:
Screenshot 2023-09-06 at 6 24 28 PM

The UI is mobile-first:

Screenshot 2023-09-06 at 6 31 17 PM

---

https://github.com/RetroAchievements/RAWeb/assets/3984985/82644715-484d-4009-b1f8-b9274b249342

---

### Implementation Details

* This page is fully built from top to bottom using Blade components. The only exception is a single call to `userAvatar()` and a few calls to `renderGameTitle()`.
* Nearly all of the data on the page is derived from a single call to `getUsersCompletedGamesAndMax()` and `getUsersSiteAwards()`. Where applicable, aggregated queries have been updated to support the imminent cutover.
* The page is currently protected by auth. If the user is not authenticated, the page returns a 401.
* _completion-progress.css_ has been added, which dramatically reduces the document size of the page.
* Links have been attached to the user profiles Progression Status component, and the counts logic is now shared between this new page and the Progression Status component.

**Acknowledgements**
The desktop design for this page was heavily inspired by PSNProfiles.

### CODE REVIEW

File: resources/views/platform/completion-progress-page.blade.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                <p class="text-xs mb-2">
+                    Viewing
+                    <span class="font-bold">{{ localized_number($totalInList) }}</span>
+                    @if ($isFiltering)
+                        of {{ $primaryCountsMetrics['numPlayed' ]}}
+                    @endif
+                    {{ trans_choice(__('resource.game.title'), $isFiltering ? $primaryCountsMetrics['numPlayed'] : $totalInList) }}
+                </p>
+            @endif
+        @elseif ($totalInList === 0)
```
Comment (Tsearo) [inline resources/views/platform/completion-progress-page.blade.php:58]: I see a message when there are no matching games for the current filter for users with no activity but I do not see that message when clicking someone that has something like 0 games beaten and therefore no matching games when clicked.

Ex. No achievements earned
image

Ex. Clicking 0 Beaten for someone with achievements earned
image

File: app/Platform/Controllers/PlayerCompletionProgressController.php [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+                $metrics['numCompleted']++;
+            } elseif ($game['HighestAwardKind'] === 'mastered') {
+                $metrics['numMastered']++;
+            }
+        }
+
+        return $metrics;
+    }
+
+    private function buildMilestones(array $filteredAndJoinedGamesList): array
```
[the remark this answers is not in the document]
Comment (wescopeland) [reply app/Platform/Controllers/PlayerCompletionProgressController.php:151]: I've added beaten softcore/hardcore to use landmarks in https://github.com/RetroAchievements/RAWeb/pull/1833/commits/2de84d321fb27c08bdf2a69abd87abe9979c3ae4. This commit also adds some basic test coverage for milestones.

File: app/Platform/Controllers/PlayerCompletionProgressController.php [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+            'any-softcore' => fn ($game) => $game['PctWon'] !== $game['PctWonHC'],
+
+            'eq-beaten-softcore' => fn ($game) => isset($game['HighestAwardKind']) && $game['HighestAwardKind'] === 'beaten-softcore',
+            'eq-beaten-hardcore' => fn ($game) => isset($game['HighestAwardKind']) && $game['HighestAwardKind'] === 'beaten-hardcore',
+            'eq-completed' => fn ($game) => isset($game['HighestAwardKind']) && $game['HighestAwardKind'] === 'completed',
+            'eq-mastered' => fn ($game) => isset($game['HighestAwardKind']) && $game['HighestAwardKind'] === 'mastered',
+
+            'eq-revised' => fn ($game) => isset($game['HighestAwardKind'])
+                && ($game['HighestAwardKind'] === 'completed' || $game['HighestAwardKind'] === 'mastered')
+                && ($game['PctWon'] < 1 || $game['PctWonHC'] < 1),
```
[the remark this answers is not in the document]
Comment (wescopeland) [reply app/Platform/Controllers/PlayerCompletionProgressController.php]: Good catch, resolved in https://github.com/RetroAchievements/RAWeb/pull/1833/commits/f4d48a7f657216ee1c39873d50c64ad720d3394b.

File: resources/views/platform/completion-progress-page.blade.php (hunk 1 above)
Comment (wescopeland) [reply resources/views/platform/completion-progress-page.blade.php]: Looks like I introduced some flawed conditional logic in a refactor, should be resolved as of https://github.com/RetroAchievements/RAWeb/pull/1833/commits/8279fc6f0cdd593e76cea210bd7eaae2e42a1465.

**Comment by wescopeland** [comment]
I've added links to the new Progression Status component on user profiles which now take the user directly to this page with the correct filters applied.

File: resources/views/community/components/user/progression-status/console-progression-list-item.blade.php [hunk 4]
```diff
@@ -28,11 +28,19 @@
 
 $displayLabel = $label ?? config('systems')[$consoleId]['name_short'];
 $consoleTooltipLabel = $label ?? config('systems')[$consoleId]['name'];
+
+$targetUser = request('user');
+$cellUrls = [
+    'totals' => route('user.completion-progress', [$targetUser, 'filter[system]' => $consoleId]),
+    'unfinished' => route('user.completion-progress', [$targetUser, 'filter[system]' => $consoleId, 'filter[status]' => 'unawarded']),
+    'any-beaten' => route('user.completion-progress', [$targetUser, 'filter[system]' => $consoleId, 'filter[status]' => 'any-beaten']),
+    'any-mastery' => route('user.completion-progress', [$targetUser, 'filter[system]' => $consoleId, 'filter[status]' => 'gte-completed']),
```
Comment (Jamiras) [inline resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: There seems to be a calculation difference between the list and the page. My user page says I have 45 NES games mastered:
image

But the list page says there's 46:
image

Comment (Jamiras) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: A similar thing happens with my GBC lists. I think it's the "Set has been retired" items that are throwing it off.

[the remark this answers is not in the document]
Comment (wescopeland) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: This came up in a conversation with @luchaos. He says the current query param approach we are taking will be replaced with JSON API-style resource grids, which use params such as:

`?filter[system]=1`

and for new greenfield development it makes sense to use this approach.

**Comment by Jamiras** [comment]
I could see this widget in the Recently Played Games section:
image

Possibly with the ability to expand/collapse each game to see the achievements

vs (current):
image

File: resources/views/platform/components/completion-progress-page/meta-panel/index.blade.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
+@props([
```
[the remark this answers is not in the document]
Comment (wescopeland) [reply resources/views/platform/components/completion-progress-page/meta-panel/index.blade.php:1]: Sorting functionality has been added in latest. There are four available sort orders: newest unlock (default), oldest unlock, most achievements won, and fewest achievements won.

Additionally, if any filter or sort is set, a new button appears letting the user easily reset back to the default view.

Screenshot 2023-09-10 at 6 15 22 PM

**Comment by wescopeland** [comment]
Good idea. In a future PR we could expand these list components to make them collapsible with the earned achievements, then bring them to user profiles in this way. They probably would look a lot better than what's there right now.

File: resources/views/community/components/user/progression-status/console-progression-list-item.blade.php (hunk 4 above)
Comment (wescopeland) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: I think there was a caching issue going on. I'm seeing correct numbers now on latest after updating the cache key:

Screenshot 2023-09-10 at 6 24 57 PM

Comment (Jamiras) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: Which cache key? I commented out this code and didn't see any change:
```
        $cachedAwardedValues = Cache::get(CacheKey::buildUserCompletedGamesCacheKey($user));
        if ($cachedAwardedValues) {
            return getLightweightUsersCompletedGamesAndMax($user, $cachedAwardedValues);
        }
```

Comment (wescopeland) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: I'm able to repro with another user- working through this now.

Comment (wescopeland) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: Confirmed that totals were not being counted up correctly and some awards were not being filtered out properly from the tallies. Fixes made in https://github.com/RetroAchievements/RAWeb/pull/1833/commits/fd4d4841ddf9c273461d4459c5f2f16dcf63732b and https://github.com/RetroAchievements/RAWeb/pull/1833/commits/d129d8ac6de5fb80a3001870714b8cea2326cded.

**Comment by PopovMark** [comment]
Would it be possible to implement a sorting by the percentage of points earned? That is, points earned in the set divided by the total points of the set. Both from highest to lowest and vice versa.

**Comment by wescopeland** [comment]
> Would it be possible to implement a sorting by the percentage of points earned?

Points are not currently available in these entities.

File: app/Platform/Services/PlayerProgressionService.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+            if (!$alreadyProcessed && !isset($processedGameIds[$gameId])) {
+                $searchResults = array_filter($siteAwards, fn ($award) => $award['AwardData'] == $gameId);
+
+                $award = null;
+                if (!empty($searchResults)) {
+                    $award = current($searchResults);
+                }
+
+                if ($award && isValidConsoleId($award['ConsoleID']) && $award['ConsoleID'] != 101) {
```
Comment (Jamiras) [inline app/Platform/Services/PlayerProgressionService.php:130]: I'm getting a casting error on this line:
```
isValidConsoleId(): Argument #1 ($consoleId) must be of type int, string given, called in /home/jamiras/source/RAWeb/app/Platform/Services/PlayerProgressionService.php on line 130
```

File: app/Platform/Controllers/PlayerCompletionProgressController.php [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+            'totalPages' => $totalPages,
+            'user' => $foundTargetUser,
+        ]);
+    }
+
+    private function useSortedList(array $filteredAndJoinedGamesList, string $sortOrder): array
+    {
+        if ($sortOrder === 'unlock_date') {
+            usort($filteredAndJoinedGamesList, function ($a, $b) {
+                return strtotime($b['MostRecentWonDate']) - strtotime($a['MostRecentWonDate']);
```
Comment (Jamiras) [inline app/Platform/Controllers/PlayerCompletionProgressController.php:130]: I apparently have awards without `MostRecentWonDate`s:
```
strtotime(): Argument #1 ($datetime) must be of type string, null given
```
I can no longer open the page for most of the consoles in my list.

File: resources/views/community/components/user/progression-status/console-progression-list-item.blade.php (hunk 4 above)
Comment (Jamiras) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: I still only see 45 in my profile page list, and clicking on it gives the NULL datetime error I mentioned above.

Comment (wescopeland) [reply resources/views/community/components/user/progression-status/console-progression-list-item.blade.php:37]: Thanks for walking through this with me. I am hoping the changes in the latest commit resolve some of the issues you're seeing locally.

**Review by Jamiras — APPROVED** [review]

File: app/Platform/Controllers/PlayerCompletionProgressController.php (hunk 7 above)
Comment (Tsearo) [inline app/Platform/Controllers/PlayerCompletionProgressController.php:130]: I am running into an error where this array key `MostRecentWonDate` (and sometimes `FirstWonDate`) are undefined for me after testing beating a game in softcore locally and having it be my last completed game

File: resources/views/platform/components/completion-progress-page/milestones/table-row.blade.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+if ($milestone['kind'] === 'mastered') {
+    $kindLabel .= ' mastery';
+} elseif ($milestone['kind'] === 'completed') {
+    $kindLabel .= ' completion';
+} elseif ($milestone['kind'] === 'beaten-hardcore') {
+    $kindLabel .= ' game beaten';
+}elseif ($milestone['kind'] === 'beaten-softcore') {
+    $kindLabel .= ' game beaten*';
```
Comment (Tsearo) [inline resources/views/platform/components/completion-progress-page/milestones/table-row.blade.php:24]: This does not appear to show a tooltip on hover for me as we discussed
image

Comment (wescopeland) [reply resources/views/platform/components/completion-progress-page/milestones/table-row.blade.php:24]: Fixed in https://github.com/RetroAchievements/RAWeb/pull/1833/commits/5521ad5f127246f292d0fa55406e04124c64e27f.

File: app/Platform/Controllers/PlayerCompletionProgressController.php (hunk 7 above)
Comment (Tsearo) [reply app/Platform/Controllers/PlayerCompletionProgressController.php:130]: Definitely seemed to be cache related after our testing but I wasn't able to reproduce it following the same steps (reset game progress then re-complete it) so hopefully your latest commit will handle that better.

Comment (Jamiras) [reply app/Platform/Controllers/PlayerCompletionProgressController.php:130]: OK. Figured it out.

With `FEATURE_AGGREGATE_QUERIES` enabled, I get back a null `MostRecentWonDate` from `pg.last_unlock_at AS MostRecentWonDate`.

With `FEATURE_AGGREGATE_QUERIES` disabled, I get back an empty string for `MostRecentWonDate` from `MAX(aw.Date) AS MostRecentWonDate`.

`strtotime` appears to work with empty string, but not with null.

Comment (wescopeland) [reply app/Platform/Controllers/PlayerCompletionProgressController.php]: Really good find, thank you.

In https://github.com/RetroAchievements/RAWeb/pull/1833/commits/907562b7b77e1ebdf28e41d42d548489aeb2bf98 I've added some nullish coalescing fallbacks:

```php
if ($sortOrder === 'unlock_date') {
    usort($filteredAndJoinedGamesList, function ($a, $b) {
        return strtotime($b['MostRecentWonDate'] ?? '') - strtotime($a['MostRecentWonDate'] ?? '');
    });
}

if ($sortOrder === '-unlock_date') {
    usort($filteredAndJoinedGamesList, function ($a, $b) {
        return strtotime($a['MostRecentWonDate'] ?? '') - strtotime($b['MostRecentWonDate'] ?? '');
    });
}
```

**Pull request merged**